adaptris/interlok-aws
# INTERLOK-3211 new interlok-aws-kms sub module

## Motivation

Create a set of services that interact with Amazon KMS that allow us to 

* Encrypt
* Decrypt
* Sign
* Verify 

In actual fact the verify service  isn't necessarily that useful because KMS doesn't allow us to use arbitrary public keys it's always a key pair, and most "verification scenarios" involve a situation where you have your partners public key (or certificate) only; and not their private key for obvious reasons.

## Modification

New sub-module interlok-aws-kms that provides the services.

## Result

New services; this is a work-in-progress, there are no encryp/decrypt services yet.

## Testing

* create yourself a CMK for the operation you want to do.
* make a note of the GUID; if using an alias then it's "alias/My Alias Name"
Your configuration will look something like this, we hash the payload before signing because KMS restricts us to 4096 bytes; the message type can be RAW, if you expect verification to be done outside of KMS; it can be DIGEST if you're using kms-verify to verify it.

```
              <payload-hashing-service>
                <unique-id>hash-payload</unique-id>
                <hash-algorithm>SHA-256</hash-algorithm>
                <metadata-key>payload-hash</metadata-key>
                <byte-translator class="base64-byte-translator"/>
              </payload-hashing-service>
              <aws-kms-generate-signature>
                <unique-id>create-signature</unique-id>
                <connection class="shared-connection">
                  <lookup-name>kms-connection</lookup-name>
                </connection>
                <key-id>55c49099-df7c-47c4-AAA-b8c8606bc923</key-id>
                <signing-algorithm>RSASSA_PSS_SHA_256</signing-algorithm>
                <message-type>RAW</message-type>
                <input class="byte-array-from-metadata">
                  <key>payload-hash</key>
                  <translator class="base64-byte-translator"/>
                </input>
                <output class="metadata-stream-output">
                  <metadata-key>signature</metadata-key>
                  <translator class="base64-byte-translator"/>
                </output>
              </aws-kms-generate-signature>
```
